Implement trash bin and bulk grain action buttons.

dwrensha opened the pull request:
This implements most of #1915.

The grain list looks like: 
![grains](https://cloud.githubusercontent.com/assets/495768/15087843/d79731a4-13ba-11e6-9ba8-dec212c94313.png)

When you click "View Trash", you get:
![trash](https://cloud.githubusercontent.com/assets/495768/15087845/e919eb24-13ba-11e6-9366-f20d413d0749.png)

The "Delete permanently" button updates its text depending on what's selected. If only "shared with me grains" are selected, it says "Forget permanently". If both kinds of grains are selected, it says "Delete/forget permanently".
![forget](https://cloud.githubusercontent.com/assets/495768/15087868/2376c79c-13bb-11e6-9de5-6d874c2399c3.png)

The app details page gets a similar update:
![app-details](https://cloud.githubusercontent.com/assets/495768/15087946/c8f667a4-13bb-11e6-8bf6-171810932064.png)


If you search for something in the main list and matches only exist in the trash list, you get a pointer like this: 
![no-match](https://cloud.githubusercontent.com/assets/495768/15087883/501119ce-13bb-11e6-8d5d-9dcd729a21d9.png)

If you try to open a grain that is in your trash, you get an error page like this:
![in-trash-interstitial](https://cloud.githubusercontent.com/assets/495768/15087900/6e36a84c-13bb-11e6-9cda-7491b7dc7760.png)

If you try to open a grain that is not owned by you and is in its owner's trash, you get an error page like this: 
![owner-trash](https://cloud.githubusercontent.com/assets/495768/15087927/a47d52ca-13bb-11e6-988d-23e6ccbf2156.png)

This also updates the permissions computation so that when a grain gets trashed or deleted, any sessions that depended on that grain are immediately closed.

ocdtrekkie commented:
I am so excited about this I can't keep it to myself, and have to tell you. :D

neynah commented:
I'm guessing Select all(list) as well as secondary buttons to "Return to Main list" (in the last 2 screenshots) are still being worked on?

Otherwise, looks great @dwrensha!

kentonv commented on shell/server/grain-server.js:
```diff
@@ -273,12 +273,75 @@ Meteor.publish("grainSize", function (grainId) {
+SandstormDb.periodicCleanup(86400000, () => {
+  const trashExpiration = new Date(Date.now() - GRAIN_DELETION_MS);
+  Grains.find({ trashed: { $lt: trashExpiration } }).forEach((grain) => {
+    Grains.remove({ _id: grain._id });
```
Maybe I'm missing something but it looks like this only removes the grain metadata from mongo, not the actual grain data from the back-end?

dwrensha replied:
Good catch! Also, the clearing of `ApiTokens` should be outside of this iteration. Fixed.

kentonv commented on shell/server/grain-server.js:
```diff
@@ -273,12 +273,75 @@ Meteor.publish("grainSize", function (grainId) {
+
+    ApiTokens.remove({ trashed: { $lt: trashExpiration } });
+
+    // Don't need to update demo stats: demo grains never get here!
```
But DeleteStats isn't just for demo users...

dwrensha replied:
Fixed.

kentonv commented:
Hmm, it occurs to me that trash could be confusing for people hitting quota limits. Maybe we need an "empty trash" button added to the "out of quota" prompt? Or maybe we should automatically remove items from the trash if the user is out of quota?

kentonv commented:
Hmm, I see that there's no "select / unselect all" checkbox at the top of the check box column. I feel like this is desirable, especially for "unselect all" -- If I select a bunch of things then change my mind, I have to painstakingly click each one to unselect it, or refresh the whole page.

On a similar note, there probably needs to be a one-click "empty trash" button in the trash. Currently if I want to clear everything (e.g. to free up quota) I have to click to select each item...

kentonv commented:
Can we make the check-box click targets bigger? Like, additionally listen for clicks on the whole table cell. When going down the list selecting things I find myself clicking in the space between two checkboxes a lot.

kentonv commented:
Bug: When confirming installation of a new app, the "view trash" button is visible but doesn't do anything.

neynah commented:
@paulproteus mentioned something about displaying "Grains: 4 (plus 2 in trash)".

I think we should have the number alongside "View trash". So "View trash (2)" for example.

dwrensha commented:
> Hmm, it occurs to me that trash could be confusing for people hitting quota limits. Maybe we need an "empty trash" button added to the "out of quota" prompt? Or maybe we should automatically remove items from the trash if the user is out of quota?

Automatically deleting things sounds scary to me. Putting a button on the prompt sounds fine. That will need to go in a separate pull request agains the blackrock repo.

dwrensha commented:
> Hmm, I see that there's no "select / unselect all" checkbox at the top of the check box column. I feel like this is desirable, especially for "unselect all" -- If I select a bunch of things then change my mind, I have to painstakingly click each one to unselect it, or refresh the whole page.

I've implemented "select all" and "unselect all". If anything is selected, the `select-all` checkbox is checked, and clicking on it unselects everything.

> On a similar note, there probably needs to be a one-click "empty trash" button in the trash. Currently if I want to clear everything (e.g. to free up quota) I have to click to select each item...

My feeling is that "select all" is sufficient to address this concern.

If we decide otherwise, where should the "empty trash" button go?

dwrensha commented:
I've addressed all comments. With the new additions, the UI looks like:

![select-all](https://cloud.githubusercontent.com/assets/495768/15126636/d130759e-1600-11e6-8bc3-a55754c6bd96.png)

ocdtrekkie commented:
Select All + Delete Permanently = Empty Trash seems reasonable. I don't like "delete everything" to ever be a one click operation.

zarvox commented:
> My feeling is that "select all" is sufficient to address this concern.
> 
> If we decide otherwise, where should the "empty trash" button go?

FWIW, I favor having the "empty the trash" button, since I think it's a reasonable thing to have, and shortens the "I don't always delete things, but when I do, I want them gone" workflow.  I expect that many users will only visit the trash list to reduce their quota impact, and the best thing for them is to just finish deleting the things they already expressed intent to delete, rather than making them figure out multiselect, especially if they haven't bothered learning how to use that feature yet.  I believe that there's probably a decently-sized fraction of Gmail users who have never used the checkboxes/multiselect feature and work entirely from clicking threads and the buttons in the button row; this affordance would serve that class of users well.  (These users also benefit from retaining the trash button in the viewing-a-grain topbar view - they can navigate to the item, and then take the action they expect to be able to take on it.)

I'd support a separate "empty the trash" small, link-styled button in the "Grains in the trash will be permanently removed after 30 days." paragraph, like how Gmail adds the option when you have items in the trash.  Perhaps:

```html
<p>
  <a onclick="emptyTrash();">Empty trash now</a> (Grains that have been in the trash for more than 30 days will be removed permanently.  To keep a grain, restore it to the grain list.)
</p>
```

> I don't like "delete everything" to ever be a one click operation.

To be fair, with this set of changes, It's not -- you have to send everything to the trash first, and then empty the trash.  The data there is already going to expire if you leave it alone; this is an affordance to allow the user to converge on that state sooner because the

neynah commented:
I'm +1 on the "empty all trash" link as long as there's a prompt to confirm that you really do want to delete all the (#) grains. This behaviour is very common & expected. Also, people misclick things all the time. :)

dwrensha commented:
I've implemented zarvox's suggestion:

![emptytrash](https://cloud.githubusercontent.com/assets/495768/15131411/a55d5dfc-161d-11e6-9aba-3f15c0b79b19.png)

When you click, you get a popup confirmation:

![emptytrash-popup](https://cloud.githubusercontent.com/assets/495768/15131413/a8332570-161d-11e6-9990-90febab6ddac.png)

zarvox commented:
:100: 

Someday (not today) I hope to replace all of our uses of `window.confirm()` with app-modal dialogs like those used on the (new) identity provider settings page.

dwrensha commented:
One more update: now the trash bin has the route "/grain#trash". This makes it possible to link directly to it, which should make it easy to implement @neynah's suggestion in https://github.com/sandstorm-io/sandstorm/issues/1915#issuecomment-217980290.

kentonv commented on shell/server/grain-server.js:
```diff
@@ -273,12 +273,83 @@ Meteor.publish("grainSize", function (grainId) {
+      });
+    }
+
+    waitPromise(globalBackend.deleteGrain(grain._id, grain.userId));
```
For robustness, this call should probably happen before removing the table entry. Otherwise, a badly-timed crash could lead to the grain storage still existing with no way to clean it up.

Probably we should make the same change for the explicit delete flow, too...

dwrensha replied:
Done. I've also verified that if the grain is gone on the filesystem, the rest of this method still executes as intended. I'm not sure about Blackrock, though.

kentonv replied:
Yes, the method was intentionally designed to be idempotent for this reason.

kentonv commented on shell/packages/sandstorm-ui-grainlist/grainlist-client.js:
```diff
@@ -357,6 +357,14 @@ Template.sandstormGrainTable.events({
     }
   },
 
+  "click td.select-grain": function (event, instance) {
```
This doesn't seem to have worked AFAICT: clicking just outside the checkbox does not check it for me.

kentonv replied:
Wait, never mind. I was running a different version of the code than I thought I was.

kentonv commented on shell/packages/sandstorm-permissions/permissions.js:
```diff
@@ -1203,54 +1222,71 @@ SandstormPermissions.grainPermissions = function (db, vertex, viewInfo, onInvali
+      observeHandle.push(tokenCursor.observe({
+        changed(newApiToken, oldApiToken) {
+          if (newApiToken.trashed ||
+              !_.isEqual(newApiToken.roleAssignment, oldApiToken.roaleAssignment) ||
```
We really need to switch to a type-safe language...
